### PR TITLE
Exposure settings ability for Categories and Challenges

### PR DESCRIPTION
**Pull request by Webmast84**
Added the ability to select if a challenge or category individually should be displayed or not. These options have been added to the admin pages and reflect on the challenges page.

### CODE REVIEW

File: htdocs/admin/new_category.php [hunk 1]
```diff
@@ -13,6 +13,7 @@
 form_textarea('Description');
 form_input_text('Available from', date_time());
 form_input_text('Available until', date_time());
+form_input_checkbox('Expose');
```
Comment (Nakiami) [inline htdocs/admin/new_category.php]: Can we do ('Expose', true) to tick box by default?

File: htdocs/admin/new_challenge.php [hunk 2]
```diff
@@ -15,6 +15,7 @@
 form_textarea('Flag');
 form_input_checkbox('Automark', true);
 form_input_checkbox('Case insensitive');
+form_input_checkbox('Expose');
```
Comment (Nakiami) [inline htdocs/admin/new_challenge.php]: Can we do ('Expose', true) to tick box by default?

File: htdocs/challenges.php [hunk 3]
```diff
@@ -83,11 +87,12 @@
         echo '<li ',($current_category['id'] == $cat['id'] ? ' class="active"' : ''),'><a href="',CONFIG_SITE_URL,'challenges?category=',htmlspecialchars($cat['id']),'">',htmlspecialchars($cat['title']),'</a></li>';
     }
 }
+}
 echo '</ul>
 </div>';
 
 // check that the category is actually available for display
-if ($time < $current_category['available_from'] || $time > $current_category['available_until']) {
+if ($time < $current_category['available_from'] || $time > $current_category['available_until'] || $current_category['expose'] ==0) {
```
Comment (Nakiami) [inline htdocs/challenges.php]: Can change ``$current_category['expose'] ==0`` to `!$current_category['expose']`

File: htdocs/challenges.php [hunk 4]
```diff
@@ -130,6 +136,7 @@
 
     // if the challenge isn't available yet, display a message and continue to next challenge
     if ($time < $challenge['available_from']) {
+        if($challenge['expose'] ==1) {
```
Comment (Nakiami) [inline htdocs/challenges.php]: These two lines could be
``
if ($time < $challenge['available_from'] && $challenge['expose']) {
``

File: htdocs/challenges.php [hunk 5]
```diff
@@ -153,12 +161,14 @@
     } else if ($challenge['correct_submission_added']) {
         $panel_class = "panel-success";
     }
+    if($time < $challenge['available_from'] && $challenge['expose'] ==1 || $time >= $challenge['available_from']) {
```
Comment (Nakiami) [inline htdocs/challenges.php]: ``$challenge['expose'] ==1`` can be ``$challenge['expose']``

**Comment by Nakiami** [comment]
Thanks, looks like a good improvement! Maybe you could add indicators to the main overview page, so you can get a quick overview of which categories and challenges are exposed?

**Comment by Webmast84** [comment]
Sounds good. I will make these changes later today.

**Comment by Webmast84** [comment]
Moved exposure checking logic into MySQL, added admin quick indicator icons for categories and challenges

File: htdocs/admin/index.php [hunk 6]
```diff
@@ -66,6 +68,7 @@
           <td>
             <a href="edit_challenge.php?id=', htmlspecialchars($challenge['id']), '" class="btn btn-xs btn-primary">Edit</a>
             <a href="new_hint.php?id=', htmlspecialchars($challenge['id']), '" class="btn btn-xs btn-warning">Hint</a>
+            <img src="/img/', ((int)$challenge['expose'] ==1 ? 'accept.png" alt="Exposed"' : 'stop.png" alt="Not Exposed"'), ' />
```
Comment (Nakiami) [inline htdocs/admin/index.php]: ``(int)$challenge['expose'] ==1`` could be ``$challenge['expose']``

Comment (Webmast84) [inline htdocs/admin/index.php]: Our environment doesn't allow for comparing integer results as boolean. When we tried doing the example you suggested in our setup the result comes back as 0 or 1 and comparing as boolean doesn't work. That's also the reason I removed all those other checks in PHP and just added it to the SQL query.

Comment (Nakiami) [inline htdocs/admin/index.php]: Really? What's your setup?

Comment (jrespeto) [inline htdocs/admin/index.php]: I added and retested the requested change, looks to work well. 

changed
    (int)$challenge['expose'] ==1

    to
     (int)$challenge['expose']

**Comment by jrespeto** [comment]
Hi Nakiami; can this get merged? 
Thanks

**Comment by Nakiami** [comment]
Functionality added as part of https://github.com/Nakiami/mellivora/commit/87bec9fed251cc5879c0fa76b696902a297bdffd

**Pull request closed without merging**